File: parser/src/PB/AST/Expr.hs
module PB.AST.Expr
  ( LvSegment (..)
  , Lvalue (..)
  , BinOp (..)
  , DispatchMode (..)
  , DispatchExpr (..)
  , Expr (..)
  ) where

import PB.Prelude
import PB.Lexing.Token (Token (..))
import GHC.Generics (Generic)

-- | One segment of a dotted name, e.g. the `arr[i]` in `obj.arr[i].field`.
-- subscript holds the raw subscript expression tokens when present.
data LvSegment = LvSegment
  { name      :: Text
  , subscript :: Maybe [Text]
  } deriving (Eq, Show, Generic)

newtype Lvalue = Lvalue
  { segments :: [LvSegment]
  } deriving (Eq, Show, Generic)

data BinOp
  = BopAdd | BopSub | BopMul | BopDiv | BopPow
  | BopEq  | BopNe  | BopLt  | BopGt  | BopLe | BopGe
  | BopAnd | BopOr  | BopXor
  deriving (Eq, Show, Generic)

data DispatchMode = DmPost | DmTrigger | DmSync
  deriving (Eq, Show, Generic)

data DispatchExpr = DispatchExpr
  { object  :: Maybe Lvalue
  , mode    :: DispatchMode
  , dynamic :: Bool
  , event   :: Bool
  , name    :: Text
  , args    :: [[Token]]
  } deriving (Eq, Show, Generic)

-- | Expression AST.
--
-- Single-argument constructors use positional syntax; Aeson serialises them
-- as @{"tag":"…","contents":…}@.  Multi-argument constructors use record
-- syntax with unique field names.  Token lists are stored as Text at
-- parse time.
data Expr
  -- Literals (positional → "contents" in JSON)
  = ExBool        Bool
  | ExInt         Text
  | ExReal        Text
  | ExStr         Text
  | ExDate        Text
  | ExTime        Text
  | ExNull
  -- Identifiers
  | ExEnum        Text              -- enum constant (without trailing '!')
  | ExLvalue      Lvalue            -- inlines Lvalue.segments into JSON
  -- Calls (record constructors with unique field names)
  | ExCall        { callee :: Lvalue,  callArgs :: [[Token]] }
  | ExMethodCall  { receiver :: Expr, method :: Text, methodArgs :: [[Token]] }
  | ExDispatch    DispatchExpr      -- inlines DispatchExpr fields into JSON
  -- Object creation
  | ExCreate      Text              -- CREATE ClassName
  | ExCreateUsing Expr              -- CREATE USING expr
  -- Compound
  | ExArray       [Expr]
  | ExBinOp       { lhs :: Expr, op :: BinOp, rhs :: Expr }
  | ExNot         Expr
  | ExNeg         Expr              -- unary minus
  -- Special
  | ExHostVar     Lvalue            -- SQL host variable :varname
  | ExRaw         [Text]            -- unrecognised / SQL fragment tokens
  deriving (Eq, Show, Generic)

File: ui/packages/interpreter/src/instr/expr.ts
// interpreter/instr/expr.ts — Expression evaluator for the InstrGraph compiler and runner.

import type { Expr } from "../types/ast.js";
import { PB_BUILTINS } from "../runtime.js";
import { type VarEnv, readVar } from "./var-env.js";

export function evalExpr(env: VarEnv, expr: Expr): unknown {
  switch (expr.tag) {
    case "ExBool":   return expr.contents;
    case "ExInt":    return parseInt(expr.contents, 10);
    case "ExReal":   return parseFloat(expr.contents);
    case "ExStr":    return expr.contents;
    case "ExDate":   return expr.contents;
    case "ExTime":   return expr.contents;
    case "ExNull":   return null;
    case "ExEnum":   return expr.contents;
    case "ExLvalue": {
      const name = expr.contents.segments[0]?.name;
      return name ? readVar(env, name) : undefined;
    }
    case "ExCall": {
      const callee = expr.callee.segments.map((s) => s.name).join(".");
      const args = expr.args.map((a) => evalExpr(env, a));
      const fn = PB_BUILTINS[callee];
      return fn ? fn(...args) : undefined;
    }
    case "ExBinOp":
      return evalBinOp(evalExpr(env, expr.lhs), expr.op, evalExpr(env, expr.rhs));
    case "ExNot":    return !evalExpr(env, expr.contents);
    case "ExNeg":    return -(evalExpr(env, expr.contents) as number);
    default:         return undefined;
  }
}

function evalBinOp(l: unknown, op: string, r: unknown): unknown {
  switch (op) {
    case "BopAdd": return (l as number) + (r as number);
    case "BopSub": return (l as number) - (r as number);
    case "BopMul": return (l as number) * (r as number);
    case "BopDiv": return (l as number) / (r as number);
    case "BopPow": return Math.pow(l as number, r as number);
    case "BopEq":  return l === r;
    case "BopNe":  return l !== r;
    case "BopLt":  return (l as number) < (r as number);
    case "BopGt":  return (l as number) > (r as number);
    case "BopLe":  return (l as number) <= (r as number);
    case "BopGe":  return (l as number) >= (r as number);
    case "BopAnd": return !!(l) && !!(r);
    case "BopOr":  return !!(l) || !!(r);
    case "BopXor": return !!(l) !== !!(r);
    default: return undefined;
  }
}
